File: leetcode/222.count-complete-tree-nodes/go/countnodes.go
package countnodes

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func countNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left := root.Left
	right := root.Right

	count := 0

	for left != nil && right != nil {
		count++
		left = left.Left
		right = right.Right
	}

	if left == nil && right == nil {
		return 2 << count - 1
	}

	return 1 + countNodes(root.Left) + countNodes(root.Right)
}

File: leetcode/222.count-complete-tree-nodes/ts/countnodes.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const countNodes = (root: TreeNode | null): number => {
    if (root === null) return 0

    let leftDepth = 0
    let rightDepth = 0

    let left = root.left
    let right = root.right

    while (left !== null) {
        leftDepth++
        left = left.left
    }

    while (right !== null) {
        rightDepth++
        right = right.right
    }

    if (leftDepth === rightDepth) return (2 << leftDepth) - 1

    return 1 + countNodes(root.left) + countNodes(root.right)
}

export default countNodes
